src/nimble_model_data_base.h

```diff
@@ -168,6 +172,36 @@ class ModelDataBase {
                                     const nimble::Viewify<2> &displacement,
                                     nimble::Viewify<2> &force) {};
 
+  /// \brief Apply initial conditions
+  virtual void ApplyInitialConditions(nimble::DataManager &data_manager);
+
+  /// \brief Apply kinematic conditions
+  virtual void ApplyKinematicConditions(nimble::DataManager &data_manager,
+                                        double time_current,
+                                        double time_previous);
+
+  /// \brief Update model with new velocity
+  ///
+  /// \param[in] data_manager Reference to the data manager
+  /// \param[in] dt Current time step
+  ///
+  /// \note This routine is usually empty.
+  ///       The UQ model data is one case using this routine.
+  virtual void UpdateWithNewVelocity(nimble::DataManager &data_manager,
```

Comment:
the UQ code that you haven't seen is just using UpdateDisplacement/UpdateVelocity as hooks to trigger the two parts of the Verlet integrator i.e. these aren't triggering BCs application; however, the UQ code does need to have the BCs applied to the multiple displacements and velocities it allocates. If the model data manager holds these it's fine with me for it to also apply the boundary conditions. I'm trying to get Dave and Hemanth's ok to bring the bit of UQ code into github so this will be transparent.